### PR TITLE
Remove use of json gem

### LINKED ISSUE #2471: phased-restart errors after upgrading from Puma 3.12.6 to 5.0.4
**Describe the bug**

This seems to happen when doing phased-restarts using versioned deploy folders with symlinks.

We use a folder structure similar to this:

```
/app/app.d/{version (int)}
/app/app -> /app/app.d/{version} # symlink
/app/prev -> /app/app.d/{version-1} # symlink
```

When we deploy, we increment the version, run bundle etc, then switch the symlink and do a phased-restart of puma.

All seems to work fine, but after a few deploys, we also start pruning older versions, and then we see this error in the puma.log

```
/usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/spec_set.rb:86:in `block in materialize': Could not find json-2.3.1 in any of the sources (Bundler::GemNotFound)
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/spec_set.rb:80:in `map!'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/spec_set.rb:80:in `materialize'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/definition.rb:170:in `specs'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/definition.rb:237:in `specs_for'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/definition.rb:226:in `requested_specs'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:101:in `block in definition_method'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:20:in `setup'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler.rb:149:in `setup'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/setup.rb:20:in `block in <top (required)>'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/ui/shell.rb:136:in `with_level'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/ui/shell.rb:88:in `silence'
        from /usr/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/setup.rb:20:in `<top (required)>'
        from /usr/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:92:in `require'
        from /usr/lib/ruby/site_ruby/2.6.0/rubygems/core_ext/kernel_require.rb:92:in `require'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/configuration.rb:312:in `rack_builder'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/configuration.rb:331:in `load_rackup'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/configuration.rb:257:in `app'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/runner.rb:142:in `app'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/runner.rb:146:in `start_server'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster/worker.rb:57:in `run'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:195:in `worker'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:96:in `block in spawn_worker'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:96:in `fork'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:96:in `spawn_worker'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:78:in `block in spawn_workers'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:71:in `times'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:71:in `spawn_workers'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:142:in `check_workers'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cluster.rb:410:in `run'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/launcher.rb:171:in `run'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/lib/puma/cli.rb:80:in `run'
        from /var/local/app/app.d/1386/vendor/bundle/ruby/2.6.0/gems/puma-5.0.4/bin/puma-wild:25:in `<main>'
[12769] + Gemfile in context: /var/local/app/app.d/1391/Gemfile
[12769] Early termination of worker
```

❗️ Note that the app now lives in `/app/app.d/1390` but the errors are pointing to puma in `/app/app.d/1386` ❗️ 

**Puma config:**

```
#!/usr/bin/env puma

# The directory to operate out of.
#
# The default is the current directory.
#
directory "/var/local/app/app"

# Set the environment in which the rack's app will run. The value must be a string.
#
# The default is "development".
#
environment ENV["RAILS_ENV"] || "development"

# Store the pid of the server in the file at "path".
#
pidfile "/var/run/puma/puma.pid"

# Use "path" as the file to store the server info state. This is
# used by "pumactl" to query and control the server.
#
state_path "/var/run/puma/puma.state"

# Redirect STDOUT and STDERR to files specified. The 3rd parameter
# ("append") specifies whether the output is appended, the default is
# "false".
#
stdout_redirect "/var/log/puma.log", "/var/log/puma.log", true

# Configure "min" to be the minimum number of threads to use to answer
# requests and "max" the maximum.
#
# The default is "0, 16".
#
# threads 0, 16
threads 8, 8


# Bind the server to "url". "tcp://", "unix://" and "ssl://" are the only
# accepted protocols.
#
# The default is "tcp://0.0.0.0:9292".
#
bind 'unix:///var/run/puma/puma.sock'


# === Cluster mode ===

# How many worker processes to run.  Typically this is set to
# to the number of available cores.
#
# The default is "0".
#
workers ENV["PUMA_WORKERS"] || 16

# Allow workers to reload bundler context when master process is issued
# a USR1 signal. This allows proper reloading of gems while the master
# is preserved across a phased-restart. (incompatible with preload_app)
# (off by default)

# see https://github.com/puma/puma/blob/master/docs/deployment.md#restarting
prune_bundler
```

If I do a full restart, then things are back to normal, until it happens again (we keep 3 deploys back, so after 4 deploys...)

**To Reproduce**

I was only able to reproduce this on our staging environment so far

**Expected behavior**
phased-restart should reload everything from the new folder... It does seem to point to the right folder, but there's some mysterious left-overs pointing to something old somewhere??

**Server:**
 - OS: Ubuntu 18.04 with systemd
 - Puma Version 5.0.4 (it doesn't happen with 3.12)

### PR DESCRIPTION
**Pull request by cjlarose**
### Description

Fixes #2471. The changes proposed in this PR in particular are described in a [comment](https://github.com/puma/puma/issues/2471#issuecomment-719834731) on the same issue. This PR introduces a new `Puma::JSON` module, which is meant to handle the specific places where Puma needs to perform JSON serialization. 

### Your checklist for this pull request

- [X] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [X] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` or `[ci skip]` to the pull request title.
- [X] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [X] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [X] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [X] I have updated the documentation accordingly.
- [X] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: lib/puma/json.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+module Puma
```
Comment (nateberkopec) [inline lib/puma/json.rb:3]: You really went above and beyond here. Better than what we had in 3.x by a mile!

**Comment by nateberkopec** [comment]
100% for the approach, will do a line by line review soon

File: History.md [hunk 2]
```diff
@@ -13,6 +13,7 @@
   * Ignore illegal (by Rack spec) response header ([#2439])
   * Close idle connections immediately on shutdown ([#2460])
   * Fix some instances of phased restart errors related to the `json` gem (#2473)
+  * Remove use of `json` gem to fixed phased restart errors (#2479)
```
Comment (rmacklin) [inline History.md:16]: Minor typo here:
```suggestion
  * Remove use of `json` gem to fix phased restart errors (#2479)
```

**Review by rmacklin — COMMENTED** [review]
This is great!

The integration tests you added should prevent the specific scenarios that currently make it impossible to upgrade/downgrade `json` during a phased restart. If I understand correctly, though, the test suite could still pass even if someone were to introduce a new usage of `json` in some other code path, right?

I'm not sure there's much we can do about that (unless we want to ban the usage of `require 'json'` in general...) but just wanted to confirm that we'd still want to be on the lookout for new usages of `json` that could cause problems with phased restarts.

File: test/test_json.rb [hunk 3]
```diff
@@ -54,4 +54,14 @@ def test_json_generates_string_for_hash_with_symbol_keys
     value = { key: 'value' }
     assert_equal '{"key":"value"}', Puma::JSON.generate(value)
   end
+
+  def test_json_generates_string_for_true
+    value = [true]
+    assert_equal '[true]', Puma::JSON.generate(value)
+  end
+
+  def test_json_generates_string_for_true
```
[the remark this answers is not in the document]
Comment (cjlarose) [reply test/test_json.rb]: Commited your suggestion in  2bef0c4

File: History.md (hunk 2 above)
Comment (cjlarose) [reply History.md]: Fixed up in 0bd5531

File: lib/puma/json.rb [hunk 4]
```diff
@@ -40,6 +40,8 @@ def serialize_value(output, value)
           serialize_string output, value
         when Integer
           output << value.to_s
+        else
+          raise SerializationError, "Unexpected value of type #{value.class}"
```
[the remark this answers is not in the document]
Comment (cjlarose) [reply lib/puma/json.rb:63]: Nice catch! 90f5141

**Comment by cjlarose** [comment]
> The integration tests you added should prevent the specific scenarios that currently make it impossible to upgrade/downgrade `json` during a phased restart. If I understand correctly, though, the test suite could still pass even if someone were to introduce a new usage of `json` in some other code path, right?
> 
> I'm not sure there's much we can do about that (unless we want to ban the usage of `require 'json'` in general...) but just wanted to confirm that we'd still want to be on the lookout for new usages of `json` that could cause problems with phased restarts.

This is true. The main value of the test suite changes for phased restarts in this PR is to prevent regressions like those introduced by #1801 and #2124. There's additionally [a test that ensures that the `JSON` module isn't loaded when the puma master process starts up](https://github.com/puma/puma/blob/19e1584a134be3ce50b8852b57876fbcc4c7fe7d/test/test_integration_cluster.rb#L279-L284). But ultimately, you're right in that if someone else comes along and adds a "dynamic" `require 'json'` to the puma master process, that could introduce a new regression. It's not clear what the best path forward would be if we wanted to ensure that the `json` gem was _never_ loaded into the puma master process to prevent all regressions of this type. My hope is that the introduction of the `Puma::JSON` module is documentation enough to make it clear to contributors and to maintainers that we shouldn't `require 'json'` again.

File: lib/puma/json.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  # be particularly full-featured or fast. It just has to handle the few places
+  # where Puma relies on JSON serialization internally.
+
+  module JSON
+    class SerializationError < StandardError; end
+
+    class << self
+      def generate(value)
+        case value
+        when Array, Hash
```
Comment (ekohl) [inline lib/puma/json.rb:25]: Why is this needed? `serialize_value` already handles all types or raises a `SerializationError` if needed.

File: lib/puma/json.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+          parts = []
```
Comment (ekohl) [inline lib/puma/json.rb:26]: Rather than an array, should it use [StringIO](https://ruby-doc.org/stdlib-2.5.1/libdoc/stringio/rdoc/StringIO.html)? Not sure if there is a (performance) benefit for one or the other.

File: lib/puma/json.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  #
+  # The implementation of JSON serialization in this module is not designed to
+  # be particularly full-featured or fast. It just has to handle the few places
+  # where Puma relies on JSON serialization internally.
+
+  module JSON
+    class SerializationError < StandardError; end
+
+    class << self
+      def generate(value)
```
Comment (ekohl) [inline lib/puma/json.rb:23]: Does it make sense to attempt using the native `JSON` module for those who don't use phased restarts? It'd be faster.

File: lib/puma/json.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        output << value.gsub(/[\\"]/, '\\' => '\\\\', '"' => '\\"')
+        output << '"'
+      end
+
+      def serialize_object_key(output, value)
+        case value
+        when Symbol
+          serialize_string output, value.to_s
+        when String
+          serialize_string output, value
```
Comment (MSP-Greg) [inline lib/puma/json.rb:78]: In the past I used inspect for strings and symbols:
```ruby
when Symbol, String
  output << value.to_s.inspect
```
or 
```ruby
when Symbol
  output << value.to_s.inspect
when String
  output << value.inspect
```
I think it does all the needed processing?

File: lib/puma/json.rb (hunk 5 above)
Comment (cjlarose) [reply lib/puma/json.rb:25]: I think my implementation was based on an understanding of JSON according to RFC 4627, which requires that a valid JSON text is either an object or any array, but it looks like that requirement was dropped in RFC 7158. I'll update the implementation.

File: lib/puma/json.rb (hunk 6 above)
Comment (cjlarose) [reply lib/puma/json.rb:26]: I considered using `stringio` since it seems like a nice way to construct a string incrementally without an explosion in extra space. I went with just a normal `Array` because even `stringio` isn't in Ruby Core and, just like `json`, users can technically change the version of `stringio` they're using in a phased restart. That said `puma` _already_ uses `stringio` internally, so I could go either way on this.

I can't imagine the performance improvement would be too significant, especially since again, this module isn't really on a hot path.

**Comment by nateberkopec** [comment]
Thank you so much for the reviews @MSP-Greg @ekohl and @rmacklin 🙇

File: lib/puma/json.rb (hunk 8 above)
Comment (cjlarose) [reply lib/puma/json.rb:78]: I'll dig deeper into this. I realized I'm not handling all of the different escape sequences correctly. Newline characters, for example, are written literally instead of `\n`.

I'll have to double check that everything that `.inspect` emits is valid JSON. If it is, I'll use it, otherwise, I'll handle the special characters in a different way.

File: lib/puma/json.rb (hunk 6 above)
Comment (cjlarose) [reply lib/puma/json.rb:26]: The reason I'd switch to `stringio` is if it would improve readability/maintainability. I'll play with it and see what it looks like.

File: lib/puma/json.rb (hunk 8 above)
Comment (MSP-Greg) [reply lib/puma/json.rb:78]: Thanks.  Your comment said
```
# The implementation of JSON serialization in this module is not designed to
# be particularly full-featured or fast. It just has to handle the few places
# where Puma relies on JSON serialization internally.
```
I suspect you know better than any of us what Puma sends to `JSON.generate` (having just reviewed it), so I'll trust your judgement as to what edge cases it should handle.

File: lib/puma/json.rb (hunk 5 above)
Comment (cjlarose) [reply lib/puma/json.rb]: Fixed in 8d05a2fc832e0a1b8e142fc0d98b3075dc1f11c0

File: lib/puma/json.rb (hunk 6 above)
Comment (cjlarose) [reply lib/puma/json.rb]: Switch to StringIO in 513aab1ea343bbc3c9e2adcc3acb8645aff64c74.

I ended up doing this because I think using the `Array` approach I was doing before stood out as unusual after thinking about it. If someone was trying to build a String incrementally while keep allocations to a minimum, a reader of the code would expect the author to have used `stringio`. Deviating from that expectation would suggest that the distinction was significant, but in this case, it isn't.

File: lib/puma/json.rb (hunk 8 above)
Comment (cjlarose) [reply lib/puma/json.rb]: Added 98b3408b4de75652aca8cfc66b08a6deee6da794 to handle control characters correctly and added 28e9d8ffe5ce7e51bced113e8c266182e58f01aa to simplify the handling of `String` and `Symbol`.

AFAICT, there's no guarantee that `.inspect` produces valid JSON (or will always produce valid JSON into the future), and adding support for control characters explicitly ended up not being so bad. I think the cost of adding the cases for the different characters is small, and it makes the implementation more closely match the language of the specification. I prefer that to having to worry about some edge case down the road where `.inspect` produces something that JSON parsers don't like.

File: test/test_json.rb [hunk 9]
```diff
@@ -0,0 +1,92 @@
+require_relative "helper"
+require "puma/json"
+
+class TestJSON < Minitest::Test
```
Comment (ekohl) [inline test/test_json.rb:4]: For all of these, does it make sense to assert `JSON.parse(Puma::JSON.generate(value)) == value`?

Comment (cjlarose) [reply test/test_json.rb:5]: Ensuring that `Puma::JSON.generate` returns strings that can by parsed by `::JSON.parse` seems useful, for sure.

Updated in b905c34c264e847abed3b36d0711c95265deea2e

File: test/test_json.rb [hunk 10]
```diff
@@ -89,4 +90,17 @@ def test_generate_raises_error_for_unexpected_value_type
     end
     assert_equal 'Unexpected value of type Regexp', ex.message
   end
+
+  private
+
+  def assert_puma_json_generates_string(expected_output, value_to_serialize, expected_roundtrip: value_to_serialize)
```
Comment (ekohl) [inline test/test_json.rb:96]: I've never seen this before. Does `value_to_serialize` refer to the value that was passed in as the second argument?

Comment (cjlarose) [reply test/test_json.rb]: It seems to work that way, but I couldn't find that behavior being explicitly supported in any official Ruby documentation, so I made it a little more clear in 0bb2df3

**Comment by cjlarose** [comment]
I think I've addressed all of the feedback so far. Lemme know if there's anything else we should do!

**Review by ekohl — APPROVED** [review]
It's a complex PR so I may have missed something, but overall I think it's a good approach.
